README.md
# sequence-generator-rust

64-bit IDs sequence generator based on the concepts outlined in Twitter Server's ID (formerly snowflake). Build on Rust

## Table of Contents

- [Installation & build](#installation)
- [Description](#description)
- [Usage](#usage)
- [Library](#Library)
- [Support](#support)
- [Contributing](#contributing)

## Installation

```sh
  git clone https://github.com/drconopoima/sequence-generator-rust.git
  cd sequence-generator-rust
  cargo build --release
```

The binary was generated under `target/release/sequence-generator-rust`

## Description

You can generate sequential IDs based on timestamp, sequence number and node/worker ID (based on Twitter snowflake):

By default this package format defines:

- the right-most 9 bits are used to store worker and/or host information (up to 512)
- subsequently, 11 bits are used to store a sequence number (up to 2048)
- the left-most, 44 bits are used to store a custom epoch with precision of 10 samples every millisecond (10^-1). That's enough to store 55 years from a custom epoch
- There are no bits left unused.
- Custom epoch is set to the beginning of current decade (2020-01-01)

## Usage

Generate a single sequence number as follows, with a worker-id set up from `.env` file (default 0):

```sh
$ cargo run \
  0: 731587959438966784
```

Generate many sequence values (`-n|--number`), provide a custom worker id (`--node-id`), and measure the time taken (`-d|--debug`):

```sh
cargo run --release -- -n 8 --node-id 505 --debug
```

```text
0: 731586108621586937
1: 731586108621587449
2: 731586108621587961
3: 731586108621588473
4: 731586108621588985
5: 731586108621589497
6: 731586108621590009
7: 731586108621590521
It took 661 nanoseconds
```

Each one of the parameters for the sequence are customizable.

By default the original Twitter snowflake format defines:

- 1 bit left unused (sign)
- 41 bits are used to store a custom epoch with millisecond precision (10^3 microseconds for 69 years from a custom epoch)
- 10 bits are used to store worker and datacenter information (up to 1024)
- 12 bits are used to store a sequence number (up to 4096)
- Uses a custom epoch of 1288834974657 or Nov 04 2010 01:42:54.

You can perfectly and easily recreate Twitter's snowflakes by passing the following command arguments.

```sh
$ cargo run --release -- -n 8 -d --unused-bits 1 --node-id-bits 10 --sequence-bits 12 --micros-ten-power 3 --custom-epoch '2010-11-04T01:42:54Z'  --node-id 128
0: 137870923482005632
1: 137870923482006656
2: 137870923482007680
3: 137870923482008704
4: 137870923482009728
5: 137870923482010752
6: 137870923482011776
7: 137870923482012800
It took 571 nanoseconds
```

The specific structure of the integers at the binary level includes:

- The left-most bits (customizable, by default none) might be unused and set to 0.
- The second group of bits store the timestamp in a custom exponential by microseconds (by default `44 bits` and sampling every `100 mcs`, equivalent to argument `--micros-ten-power 2`). You cannot customize number of bits of the timestamp directly, but by indirectly setting different values for other bit groups.
- The third group of bits store the sequence (by default `11 bits`)
- The right-most group of bits store the host/worker ID (by default `9 bits`)

You can also customize by `dotenv` file. Copy the file `.env-example` into `.env`

```sh
cp .env-example .env
```

And change the example values to your liking.

The precedence of parameters assigned through the command-line launch arguments is the highest, whichever are not assigned can be retrieved by use of a `.env` file, and if still unassigned parameters remains, then default values described above are used.

The only supported custom epoch format is `RFC-3339/ISO-8601` both as CLI argument and from the dotenv file.

Check a detailed analysis for a generated value in the [auxiliar bit structure analysis](bit_structure.md)

## Benchmarking

See [auxiliar benchmarking notes](benchmarking.md)
## Library

```rust
use std::time::UNIX_EPOCH;
use ::sequence_generator::*;

let custom_epoch = UNIX_EPOCH;  // SystemTime object representing custom epoch time. Use checked_add(Duration) for different time
let node_id_bits = 10;          // 10-bit node/worker ID
let sequence_bits = 12;         // 12-bit sequence
let unused_bits = 1;            // unused (sign) bits at the start of the ID. 1 or 0 generally
let micros_ten_power = 3;       // Operate in milliseconds (10^3 microseconds)
let node_id = 500;              // Current worker/node ID
let cooldown_ns = 1500;         // initial time in nanoseconds for exponential backoff wait after sequence is exhausted

// Generate SequenceProperties
let properties = sequence_generator::SequenceProperties::new(
        custom_epoch,
        node_id_bits,
        node_id,
        sequence_bits,
        micros_ten_power,
        unused_bits,
        cooldown_ns,
    );

// Generate an ID
let id = sequence_generator::generate_id(&properties).unwrap();
// Decode ID
// Timestamp
let timestamp_micros = sequence_generator::decode_id_unix_epoch_micros(id, &properties);
// Sequence
let sequence = sequence_generator::decode_sequence_id(id, &properties);
// Node ID
let id_node = sequence_generator::decode_node_id(id, &properties);
```

## Support

Please [open an issue](https://github.com/drconopoima/sequence-generator-rust/issues/new) for support.

## Changelog

See [changelog](CHANGELOG.md)

## Contributing

Please contribute using [Github Flow](https://guides.github.com/introduction/flow/). Create a branch, add commits, and [open a pull request](https://github.com/drconopoima/sequence-generator-rust/compare/).


CHANGELOG.md
# Changelog

## 0.4.0
* bugfix: .env-example suggested SIGN_BITS instead of UNUSED_BITS, but it didn't work, added env variable support for both and additional CLI long parameter "--sign-bits"  
* feature: Input parameter validation:
  *  --sequence-bits, --node-id-bits maximum value 16.
  *  --sequence-bits, --node-id-bits minimum value 1.
  *  --unused-bits, --sign-bits maximum value 8.
* feature: new CLI short parameters
  * --number = "-n", --micros-ten-power = "-m", --node-id-bits = "-w" (as in '[w]orker bits'), --sequence-bits = "-s", --node-id = "-i", --unused-bits = "-u"
* feature: new long parameter equivalent to --number: "--quantity", and short "-q"
* feature: input validation, output warning message and exit gracefully if a number/quantity of zero ("0") ids is requested
* feature: ERROR if conflicting pairs of parameters are given: --number & --quantity or --unused-bits & --sign-bits
* feature: fully populate help messages along with values for defaults, minimum, maximum and ranges. 
* feature: Timer for --debug option now calculates directly time per generated id alongside total running time.
* feature: changed default minimum cooldown to 200 ns (down from 1500ns). 

## 0.3.0

* Replace chrono with time ^0.3. Addresses [RUSTSEC-2020-0071](https://rustsec.org/advisories/RUSTSEC-2020-0071.html): Potential Segfault in Time Crate.
* Fix parameter precedence: Command-line values take higher precedence to .env and environment variables.
* Document detail on benchmarking procedure.


src/main.rs
use ::sequence_generator::*;
use std::convert::TryFrom;
use std::env;
use std::path::Path;
use std::process;
use std::rc::Rc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use structopt::StructOpt;
use time::{format_description::well_known::Rfc3339, OffsetDateTime};

#[derive(Debug, StructOpt)]
struct Opt {
    #[structopt(
        short = "-n",
        long = "--number",
        help = "Amount/Quantity of sequence values requested. [Default: 1]"
    )]
    number: Option<usize>,
    #[structopt(
        short = "-q",
        long = "--quantity",
        help = "Amount/Quantity of sequence values requested. [Default: 1]"
    )]
    quantity: Option<usize>,
    #[structopt(
        short = "-c",
        long = "--custom-epoch",
        help = "Custom epoch in RFC3339 format. [Default: '2020-01-01T00:00:00Z' i.e. Jan 01 2020 00:00:00 UTC]",
        env = "CUSTOM_EPOCH"
    )]
    custom_epoch: Option<String>,
    #[structopt(
        short = "-m",
        long = "--micros-ten-power",
        help = "Exponent multiplier base 10 in microseconds for timestamp. [Default: 2 (operate in tenths of milliseconds)]",
        env = "MICROS_TEN_POWER"
    )]
    micros_ten_power: Option<u8>,
    #[structopt(
        short = "-w",
        long = "--node-id-bits",
        help = "Bits used for storing worker and datacenter information. [Default: 9 (range: 0-511). Maximum: 16. Minimum: 1]",
        env = "NODE_ID_BITS"
    )]
    node_id_bits: Option<u8>,
    #[structopt(
        short = "-s",
        long = "--sequence-bits",
        help = "Bits used for contiguous sequence values. [Default: 11 (range: 0-2047). Maximum: 16. Minimum: 1]",
        env = "SEQUENCE_BITS"
    )]
    sequence_bits: Option<u8>,
    #[structopt(
        short = "-i",
        long = "--node-id",
        help = "Numerical identifier for worker and datacenter information. [Default: 0]",
        env = "NODE_ID"
    )]
    node_id: Option<u16>,
    #[structopt(
        short = "-u",
        long = "--unused-bits",
        help = "Unused (sign) bits at the left-most of the sequence ID. [Default: 0. Maximum: 8]",
        env = "UNUSED_BITS"
    )]
    unused_bits: Option<u8>,
    #[structopt(
        long = "--sign-bits",
        help = "Unused (sign) bits at the left-most of the sequence ID. [Default: 0. Maximum: 8]",
        env = "SIGN_BITS"
    )]
    sign_bits: Option<u8>,
    #[structopt(
        default_value = ".env",
        long = "--dotenv-file",
        help = "File for configuration variables. [Default: '${pwd}/.env']"
    )]
    dotenv_file: String,
    #[structopt(
        long = "--cooldown-ns",
        help = "Initial time in nanoseconds for exponential backoff wait after sequence is exhausted. [Default: 200]",
        env = "COOLDOWN_NS"
    )]
    cooldown_ns: Option<u64>,
    #[structopt(short = "-d", long = "--debug")]
    debug: bool,
}

fn main() {
    let mut args = Opt::from_args();
    let dotenv_file = &args.dotenv_file;
    if Path::new(dotenv_file).exists() {
        dotenv::from_filename(dotenv_file).unwrap_or_else(|_| {
            panic!(
                "ERROR: Could not retrieve environment variables from configuration file '{}'",
                dotenv_file
            )
        });
        for (key, value) in env::vars() {
            if key == "CUSTOM_EPOCH" && !value.is_empty() && args.custom_epoch.is_none() {
                args.custom_epoch = Some(value.parse::<String>().unwrap_or_else(|_| {panic!(
                    "ERROR: Couldn't parse value CUSTOM_EPOCH '{}' as String, invalid UTF-8 characters", value)
                }));
            }
            if key == "NODE_ID_BITS" && !value.is_empty() && args.node_id_bits.is_none() {
                args.node_id_bits = Some(value.parse::<u8>().unwrap_or_else(|_| {
                    panic!(
                    "ERROR: NODE_ID_BITS '{}' couldn't be interpreted as value between 1 and 16",
                    value
                )
                }));
            }

            if key == "SEQUENCE_BITS" && !value.is_empty() && args.sequence_bits.is_none() {
                args.sequence_bits = Some(value.parse::<u8>().unwrap_or_else(|_| {
                    panic!(
                    "ERROR: SEQUENCE_BITS '{}' couldn't be interpreted as value between 1 and 16",
                    value
                )
                }));
            }
            if key == "MICROS_TEN_POWER" && !value.is_empty() && args.micros_ten_power.is_none() {
                args.micros_ten_power = Some(value.parse::<u8>().unwrap_or_else(|_| {panic!(
                    "ERROR: MICROS_TEN_POWER '{}' couldn't be interpreted as value between 0 and 64", value)
                }));
            }
            if key == "UNUSED_BITS" && !value.is_empty() && args.unused_bits.is_none() {
                args.unused_bits = Some(value.parse::<u8>().unwrap_or_else(|_| {
                    panic!(
                        "ERROR: UNUSED_BITS '{}' couldn't be interpreted as value between 0 and 8",
                        value
                    )
                }));
            }
            if key == "SIGN_BITS" && !value.is_empty() && args.unused_bits.is_none() {
                args.unused_bits = Some(value.parse::<u8>().unwrap_or_else(|_| {
                    panic!(
                        "ERROR: SIGN_BITS '{}' couldn't be interpreted as value between 0 and 8",
                        value
                    )
                }));
            }
            if key == "COOLDOWN_NS" && !value.is_empty() && args.cooldown_ns.is_none() {
                args.cooldown_ns = Some(value.parse::<u64>().unwrap_or_else(|_| {
                    panic!(
                    "ERROR: COOLDOWN_NS '{}' couldn't be interpreted as an unsigned integer value",
                    value
                )
                }));
            }
        }
    }
    if args.quantity.is_some() && args.number.is_some() {
        panic!(
            "ERROR: Conflicting parameters. Must only specify one of either '--quantity,-q' or '--number,-n'"
        )
    }
    if args.sign_bits.is_some() && args.unused_bits.is_some() {
        panic!(
            "ERROR: Conflicting parameters. Must only specify one of either '--unused-bits,-u' or '--sign-bits'"
        )
    }
    if let Some(value) = args.sign_bits {
        if value > 8 {
            panic!(
                "ERROR: SIGN_BITS '{}' is larger than the maximum value of 8.",
                value
            )
        }
    };
    if args.sign_bits.is_some() {
        args.unused_bits = args.sign_bits;
    }
    if let Some(value) = args.unused_bits {
        if value > 8 {
            panic!(
                "ERROR: UNUSED_BITS '{}' is larger than the maximum value of 8.",
                value
            )
        }
    };
    if let Some(value) = args.sequence_bits {
        if value > 16 {
            panic!(
                "ERROR: SEQUENCE_BITS '{}' is larger than the maximum value of 16.",
                value
            )
        }
        if value == 0 {
            panic!(
                "ERROR: SEQUENCE_BITS '{}' must be larger or equal than 1.",
                value
            )
        }
    };
    if let Some(value) = args.node_id_bits {
        if value > 16 {
            panic!(
                "ERROR: NODE_ID_BITS '{}' is larger than the maximum value of 16.",
                value
            )
        }
        if value == 0 {
            panic!(
                "ERROR: NODE_ID_BITS '{}' must be larger or equal than 1.",
                value
            )
        }
    };
    if args.custom_epoch.is_none() {
        args.custom_epoch = Some("2020-01-01T00:00:00Z".to_owned());
    }

    if args.micros_ten_power.is_none() {
        args.micros_ten_power = Some(2_u8);
    }

    if args.node_id_bits.is_none() {
        args.node_id_bits = Some(9_u8);
    }

    if args.sequence_bits.is_none() {
        args.sequence_bits = Some(11_u8);
    }

    if args.node_id.is_none() {
        args.node_id = Some(0_u16);
    }

    if args.unused_bits.is_none() {
        args.unused_bits = Some(0_u8);
    }

    if args.cooldown_ns.is_none() {
        args.cooldown_ns = Some(200_u64);
    }

    if args.number.is_none() {
        if let Some(value) = args.quantity {
            args.number = Some(value)
        } else {
            args.number = Some(1_usize)
        }
    }
    if let Some(value) = args.number {
        if value == 0 {
            println!("WARNING: No ids were requested. Exiting.");
            process::exit(0x0100);
        }
    }
    let custom_epoch_millis_i128 =
        OffsetDateTime::parse(args.custom_epoch.as_ref().unwrap(), &Rfc3339)
            .unwrap_or_else(|_| {
                panic!(
                    "ERROR: Could not parse CUSTOM_EPOCH '{}' as an RFC-3339/ISO-8601 datetime.",
                    args.custom_epoch.as_ref().unwrap()
                )
            })
            .unix_timestamp_nanos()
            / 1000000;
    let custom_epoch_millis = i64::try_from(custom_epoch_millis_i128).unwrap();
    let custom_epoch = UNIX_EPOCH
        .checked_add(Duration::from_millis(custom_epoch_millis as u64))
        .unwrap_or_else(|| {
            panic!(
            "ERROR: Could not generate a SystemTime custom epoch from milliseconds timestamp '{}'",
            custom_epoch_millis
        )
        });
    let properties = Rc::new(sequence_generator::SequenceProperties::new(
        custom_epoch,
        args.node_id_bits.unwrap(),
        args.node_id.unwrap(),
        args.sequence_bits.unwrap(),
        args.micros_ten_power.unwrap(),
        args.unused_bits.unwrap(),
        args.cooldown_ns.unwrap(),
    ));
    let mut vector_ids: Vec<u64> = vec![0; args.number.unwrap()];
    if args.debug {
        let time_now = SystemTime::now();
        for element in vector_ids.iter_mut() {
            *element = sequence_generator::generate_id(&properties).unwrap_or_else(
                |error| {
                    panic!(
                        "SequenceGeneratorError: Failed to get ID from properties {:?}. SystemTimeError difference {:?}",
                        properties,
                        (error).duration()
                    )
                }
            );
        }
        let elapsed = time_now
            .elapsed()
            .expect("ERROR: Failed to get elapsed time.")
            .as_nanos();
        for (index, element) in vector_ids.into_iter().enumerate() {
            println!("{}: {}", index, element);
        }
        println!(
            "It took {} nanoseconds, time per id: {:.2} ns",
            elapsed,
            elapsed as f64 / args.number.unwrap() as f64
        );
    } else {
        for (index, element) in vector_ids.iter_mut().enumerate() {
            *element = sequence_generator::generate_id(&properties).unwrap_or_else(
                |error| {
                    panic!(
                        "SequenceGeneratorError: Failed to get ID from properties {:?}. SystemTimeError difference {:?}",
                        properties,
                        (error).duration()
                    )
                });
            println!("{}: {}", index, element);
        }
    }
}
